opencti-platform/opencti-graphql/tests/01-unit/utils/confidence-level-test.ts

```diff
@@ -0,0 +1,150 @@
+    expect(cropMaxConfidenceInEditValue(-300, '/user_confidence_level/max_confidence')).toEqual(0);
+
+    // overrides.max_confidence only
+    expect(cropMaxConfidenceInEditValue(50, '/group_confidence_level/overrides/0/max_confidence')).toEqual(50);
+    expect(cropMaxConfidenceInEditValue(300, '/group_confidence_level/overrides/78/max_confidence')).toEqual(100);
+    expect(cropMaxConfidenceInEditValue(-300, '/group_confidence_level/overrides/9/max_confidence')).toEqual(0);
+    expect(cropMaxConfidenceInEditValue(50, '/user_confidence_level/overrides/8/max_confidence')).toEqual(50);
+    expect(cropMaxConfidenceInEditValue(300, '/user_confidence_level/overrides/0/max_confidence')).toEqual(100);
+    expect(cropMaxConfidenceInEditValue(-300, '/user_confidence_leve/overrides/41/max_confidence')).toEqual(0);
+
+    // throw error properly if patch is incorrect
+    expect(() => cropMaxConfidenceInEditValue(-300, '/user_confidence_leve/wrongkey/41/max_confidence'))
+      .toThrowError(FunctionalError(
+        'Unhandled object_path for patching a confidence level',
+        { object_path: '/user_confidence_level/wrongkey/41/max_confidence', value: -300 }
+      ));
+    expect(() => cropMaxConfidenceInEditValue({ entity_type: 'Report', max_confidence: 50 }, '/group_confidence_level/overrides/0/max_confidence'))
+      .toThrowError(FunctionalError(
+        'Cannot crop non-finite input value',
+        { value: { entity_type: 'Report', max_confidence: 50 } }
+      ));
+  });
+
+  it('computeUserEffectiveConfidenceLevel should correctly compute the effective level', async () => {
+    const group70 = {
+      id: 'group70',
+      group_confidence_level: {
+        max_confidence: 70,
+        overrides: [],
+      }
+    };
+
+    const group80 = {
+      id: 'group80',
+      group_confidence_level: {
+        max_confidence: 70,
+        overrides: [],
+      }
+    };
+
+    const groupNull = {
+      id: 'groupNull',
+      group_confidence_level: null
+    };
+
+    // minimal subset of a real User
+    const userA = {
+      id: 'userA',
+      user_confidence_level: {
+        max_confidence: 30,
+        overrides: [],
+      },
+      groups: [group70, group80],
+    };
+    expect(computeUserEffectiveConfidenceLevel(userA as unknown as AuthUser)).toEqual({
+      max_confidence: 30,
+      overrides: [],
+      source: userA,
+    });
+
+    const userB = {
+      id: 'userB',
+      user_confidence_level: null,
+      groups: [group70, group80],
+    };
+    expect(computeUserEffectiveConfidenceLevel(userB as unknown as AuthUser)).toEqual({
+      max_confidence: 70,
+      overrides: [],
+      source: group70,
+    });
+
+    const userC = {
+      user_confidence_level: null,
+      groups: [groupNull, group70, groupNull],
+    };
+    expect(computeUserEffectiveConfidenceLevel(userC as unknown as AuthUser)).toEqual({
+      max_confidence: 70,
+      overrides: [],
+      source: group70,
+    });
```

Comment:
This case was not handled properly before this PR. If a user had several groups, and the last one had null confidence, it would end up null.